redis_shard.go

```diff
@@ -116,36 +111,33 @@ func NewRedisShard(n *Node, conf RedisShardConfig) (*RedisShard, error) {
 		reloadPipelineCh: make(chan struct{}),
 		closeCh:          make(chan struct{}),
 	}
-	pool, err := newPool(shard, n, conf)
+	client, err := newRedisClient(shard, n, conf)
 	if err != nil {
 		return nil, err
 	}
 	shard.scripts = []*redis.Script{}
-	shard.pool = pool
+	shard.client = client
 	shard.subCh = make(chan subRequest)
 	shard.pubCh = make(chan pubRequest)
 	shard.dataCh = make(chan *dataRequest)
-	if !shard.useCluster {
```

Comment:
go-redis implements a way to run pipeline in cluster mode too, let's try it out and see whether it works。

Reply:
What is this way btw? So currently in runDataPipeline we will receive requests with keys belonging to different key slots. So we will construct a pipeline object from them all. Does go-redis do some sort of remapping in this case?

Reply:
> Does go-redis do some sort of remapping in this case?

That's right, go-redis maps commands in a pipeline to different nodes, in `ClusterClient. _processPipeline`: 

https://github.com/go-redis/redis/blob/c561f3ca7e5cf44ce1f1d3ef30f4a10a9c674c8a/cluster.go#L1062

Reply:
Hmm 🤔 While this leads to a better performance, it makes me think that go-redis approach with splitting commands by node and attempting to retry the entire pipeline in case of an error on one node may increase the chance of getting duplicate command invocations. For example:

1. We have 2 "publish with history" requests: first to node A, second to node B
2. Assuming A is currently not available, like in the process of failover
3. Part of the pipeline will be successfully executed on B, but we will get an error since it was failed on A
4. All commands in batch will get an error on Centrifuge level, possibly will be re-issued by some application code - thus there is a chance to execute publish on B again.

Think the scenario above only has a negative effect for publishing with history at the moment - i.e. when double invocations as a result of retry on application level result into duplicate messages sent and duplicate messages saved into a stream.

It seems to me that a better way to manage this is only retrying failed parts of pipeline, not the entire one. I think it's properly done in rueidis library which has implicit pipelining and we have a granular error management for pieces of pipeline out of the box. For `go-redis` I don't think we can simply use pipelining in Cluster mode given the concerns above ^

Reply:
> attempting to retry the entire pipeline in case of an error on one node

This is not right I think, `_processPipeline` records failed commands in `failedCmds`, and then attempts to run those failed only IIUIC: https://github.com/go-redis/redis/blob/c561f3ca7e5cf44ce1f1d3ef30f4a10a9c674c8a/cluster.go#L1104

> possibly will be re-issued by some application code - thus there is a good chance to execute publish on B again

This is possible, but I don't see it will happen in centrifuge. Because we collects commands and run in a pipeline in the redis_shard level, if failed, by no means we will retry the whole batch in this level (since it is dangerous), and it the upper level, it know nothing about the the pipeling thing, what it can do is just reissuing the single failed command it issued before. So I don't think the issue will happpen.

Reply:
You are right, but I meant sth different. Imagine a Redis Cluster scenario, in this case we only have one Redis shard on Centrifuge level that operates with Cluster. Now, let's suppose we have the case I described above:

1. We have 2 "publish with history" requests: first to node A, second to node B. They are combined into one pipeline and executed together inside `runDataPipeline`.
2. Assuming A is currently not available, like in the process of failover. And B is healthy
3. `go-redis` will execute part of the pipeline successfully (on B), part of pipeline with error (on A). But we will get an error from `pipe.Exec` https://github.com/centrifugal/centrifuge/blob/400bdb955c1ab8c4b797132af7a2b18352a04024/redis_shard.go#L348 since it was failed on A - because `go-redis` returns first observed error: https://github.com/go-redis/redis/blob/c561f3ca7e5cf44ce1f1d3ef30f4a10a9c674c8a/cluster.go#L1107 
4. All commands in batch will get an error on Centrifuge level due to this one error on A - see https://github.com/centrifugal/centrifuge/blob/400bdb955c1ab8c4b797132af7a2b18352a04024/redis_shard.go#L352   - while some commands in batch may be successful.

Does it make sense? Or I am missing sth? Possibly we can change Centrifuge code a bit to iterate over pipeline Exec result and manually extracting which commands were successful and which failed. And extracting cmd results from successful only.